File: 3. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        tmpMap = {}
        for left_cursor, ch in enumerate(s):
            tmpMap.clear()
            for right_cursor, c in enumerate(s):
                if right_cursor >= left_cursor:
                    # Note, if the substring without repeating chars is end with tail,
                    # then we need take care of that
                    if c in tmpMap or right_cursor == len(s):
                        break
                    # in fact, here we can store anything in tmpMap, we don't care the value
                    tmpMap[s[right_cursor]] = right_cursor
            print("length : {0}, dict is {1}".format(len(tmpMap), tmpMap))
            max_length = max(max_length, len(tmpMap))
        return max_length


def main():
    s = Solution()
    print(s.lengthOfLongestSubstring("asfdgdsfgsdffdgergsdfsadfwefdsacv"))


if __name__ == "__main__":
    main()

File: 3. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	tmpChar := []byte(s)
	fmt.Println(tmpChar)
	max_length := 0

	for left_cursor := 0; left_cursor < len(s); left_cursor++ {
		tmpMap := make(map[byte]int)
		for right_cursor := 0; right_cursor < len(s); right_cursor++ {
			if _, ok := tmpMap[s[right_cursor]]; ok || right_cursor == len(s)-1 {
				break
			}
			tmpMap[s[right_cursor]] = right_cursor
		}
		fmt.Printf("length %d, map is %+v \n", len(tmpMap), tmpMap)
		max_length = max(max_length, len(tmpMap))
	}
	return max_length

}

func max(x int, y int) int {
	if x >= y {
		return x
	}
	return y
}

func main() {
	s := "asfdgdsfgsdffdgergsdfsadfwefdsacv"
	rest := lengthOfLongestSubstring(s)
	fmt.Println(rest)
}
